### PR TITLE
let dmd.root.speller.spellerY use a seed array

### PR DESCRIPTION
**Pull request by Herringway**
Now we're getting to the good stuff. Null termination checks are no longer needed here. We can also stop copying the terminating null constantly. Next is `spellerX()`.

### CODE REVIEW

File: src/dmd/root/speller.d [hunk 1]
```diff
@@ -43,18 +43,17 @@ private bool combineSpellerResult(T)(ref T p, ref int cost, T np, int ncost)
...
-    char* buf;
-    if (seedlen <= tmp.sizeof - 2)
-        buf = tmp.ptr;
+    char[] buf;
+    if (seed.length <= tmp.sizeof - 1)
+        buf = tmp;
     else
     {
-        buf = cast(char*)alloca(seedlen + 2); // leave space for extra char
+        buf = new char[](seed.length + 1); // leave space for extra char
```
Comment (Geod24) [inline src/dmd/root/speller.d]: Why the change from alloca to GC ?

File: src/dmd/root/speller.d [hunk 2]
```diff
@@ -63,39 +62,36 @@ private auto spellerY(alias dg)(const(char)* seed, size_t seedlen, size_t index,
     searchFunctionType!dg p = null;
     int ncost;
     /* Delete at seed[index] */
-    if (index < seedlen)
+    if (index < seed.length)
     {
-        buf[index .. seedlen] = seed[index + 1 .. seedlen + 1];
-        assert(buf[seedlen - 1] == 0);
-        auto np = dg(buf[0 .. seedlen - 1], ncost);
+        buf[index .. seed.length - 1] = seed[index + 1 .. seed.length];
```
Comment (Geod24) [inline src/dmd/root/speller.d]: RHS can use `$` instead of `seed.length`

File: src/dmd/root/speller.d [hunk 3]
```diff
@@ -43,59 +43,53 @@ private bool combineSpellerResult(T)(ref T p, ref int cost, T np, int ncost)
...
+        auto np = dg(buf[0 .. seed.length - 1], ncost);
         if (combineSpellerResult(p, cost, np, ncost))
             return p;
     }
     /* Substitutions */
-    if (index < seedlen)
+    if (index < seed.length)
     {
-        buf[0 .. seedlen + 1] = seed[0 .. seedlen + 1];
+        buf[0 .. seed.length] = seed[0 .. seed.length];
```
Comment (Geod24) [inline src/dmd/root/speller.d]: RHS can be `seed`

File: src/dmd/root/speller.d [hunk 4]
```diff
@@ -43,59 +43,53 @@ private bool combineSpellerResult(T)(ref T p, ref int cost, T np, int ncost)
...
-            auto np = dg(buf[0 .. seedlen], ncost);
+            auto np = dg(buf[0 .. seed.length], ncost);
             if (combineSpellerResult(p, cost, np, ncost))
                 return p;
         }
-        assert(buf[seedlen] == 0);
     }
     /* Insertions */
-    buf[index + 1 .. seedlen + 2] = seed[index .. seedlen + 1];
+    buf[index + 1 .. seed.length + 1] = seed[index .. seed.length];
```
Comment (Geod24) [inline src/dmd/root/speller.d]: RHS == `seed`

**Review by Geod24 — DISMISSED** [review]

File: src/dmd/root/speller.d (hunk 4 above)
Comment (Herringway) [reply src/dmd/root/speller.d]: Not here, it's not starting at 0. But I switched it to $ instead of seed.length anyway.

File: src/dmd/root/speller.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/root/speller.d]: Ping on this

Comment (Herringway) [reply src/dmd/root/speller.d]: Sorry about the delay. It was a little late when I had finished that and must've misread the function as malloc or something like that.

Comment (Herringway) [reply src/dmd/root/speller.d]: Although, thinking about it, this alloca would be the only thing keeping this function from being inferred as `@safe` (when the dg parameter is also `@safe`, of course)

Comment (Geod24) [reply src/dmd/root/speller.d]: We're still quite far away from having DMD `@safe`. I think we can leave this for future improvements.

**Comment by Herringway** [comment]
I think that covers everything.
